src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -54,16 +49,34 @@ module.exports = function<T, P, I>(config : HostConfig<T, P, I>) : Reconciler {
 
   let nextUnitOfWork : ?Fiber = null;
 
-  function performUnitOfWork(unit : Fiber) : ?Fiber {
-    switch (unit.tag) {
-      case FunctionalComponent:
-        return ReactFiberFunctionalComponent.performWork(unit);
-      case ClassComponent:
-        break;
-      case HostComponent:
-        break;
+  function completeUnitOfWork(unitOfWork : Fiber) : ?Fiber {
+    while (true) {
+      var next = completeWork(unitOfWork);
+      if (next) {
+        // If completing this work spawned new work, do that next.
+        return next;
+      } else if (unitOfWork.sibling) {
+        // If there is more work to do in this parent, do that next.
+        return unitOfWork.sibling;
+      } else if (unitOfWork.parent) {
+        // If there's no more work in this parent. Complete the parent.
+        unitOfWork = unitOfWork.parent;
+      } else {
+        // If we're at the root, there's no more work to do.
+        return null;
+      }
+    }
+  }
+
+  function performUnitOfWork(unitOfWork : Fiber) : ?Fiber {
+    var next = beginWork(unitOfWork);
```

Comment:
Can you help me understand the difference between the input and output types for beginWork? They are Fiber and ?Fiber, but it seems that only the return type makes sense to pass to completeUnitOfWork. If we can't distinguish these at the actual type level, can we add a (dev-only) field in the fiber for its current state and assert what state the unit of work is in? I think that would make this easier to understand.